src/rpc/iterator/ExtGenFilm.h

```diff
@@ -0,0 +1,157 @@
+*
+* Copyright 2016 - 2022, The Regents of the University of Minnesota
+* Distributed under the terms of the GNU General Public License.
+*/
+
+#include <prdc/iterator/ExtGenFilmBase.h>  // Base class
+#include <rpc/System.h>
+
+namespace Pscf {
+namespace Rpc {
+
+   using namespace Util;
+   using namespace Pscf::Prdc;
+
+   /**
+   * External field generator for a thin film geometry.
+   *
+   * The parent ExtGenFilmBase class template defines all traits of a 
+   * ExtGenFilm that do not require access to the System. This subclass
+   * defines all methods that need System access.
+   * 
+   * If the user chooses an ExtGenFilm object to generate external fields,
+   * the external fields will have the same shape as the mask, with a 
+   * magnitude defined by a Flory--Huggins-like chi parameter. This class
+   * is specific to thin-film systems because it also allows for a 
+   * different chi parameter to be defined on the top boundary than on
+   * the bottom, through user input arrays chi_bottom and chi_top. See 
+   * \ref user_thin_films_page for more information. 
+   *
+   * \ingroup Rpc_Iterator_Module
+   */
+   template <int D>
+   class ExtGenFilm : public ExtGenFilmBase<D>
+   {
+
+   public:
+
+      /**
+      * Default constructor
+      */
+      ExtGenFilm();
+      
+      /**
+      * Constructor
+      * 
+      * \param itr  Iterator parent object
+      */
+      ExtGenFilm(System<D>& sys);
+
+      /**
+      * Destructor
+      */
+      ~ExtGenFilm();
+
+      /**
+      * Check whether the fields have been generated
+      */
+      bool isGenerated() const;
+
+      using ExtGenFilmBase<D>::isAthermal;
+      using ExtGenFilmBase<D>::chiBottom;
+      using ExtGenFilmBase<D>::chiTop;
+      using ParamComposite::setClassName;
+
+   protected:
+
+      /**
+      * Allocate container necessary to generate and store fields
+      */ 
+      void allocate();
+
+      /**
+      * Generate the fields and store where the Iterator can access.
+      */
+      void generate();
+
+      /**
+      * Get the System associated with this object by reference.
+      */
+      System<D>& system();
+      /**
+      * Get the System associated with this object by const reference.
+      */
+      System<D> const & system() const;
+
+      /**
+      * Get the space group name for this system.
+      */
+      std::string getSpaceGroup() const;
+
+      /**
+      * Get the lattice parameters for this system.
+      */
+      FSArray<double, 6> getLatticeParameters() const;
+
+      /**
+      * Get the number of monomer species for this system.
+      */
+      int getNMonomer() const;
+
+      /**
+      * Use the mask to determine the value of normalVecId
+      */
+      void getNormalVecId();
+
+   private:
+
+      /// Pointer to the associated system object.
+      System<D>* sysPtr_;
+
+      using ExtGenFilmBase<D>::parameters_;
+      using ExtGenFilmBase<D>::chiBottomCurrent_;
```

Comment:
The use of a using statement in a private block seems a little strange. This is presumably taking a variable that is protected in base class and making it private in subclasses, but why?  This is related to my question about why you have, e.g., chiBottomCurrent_ as a protected variable and chiBottom_ as private, rather than a single protected chiBottom_, so i clearly am not understanding your motivation for something, and I'll just point out that it looks strange and ask you to think about whether there is a better way or not.